2091 Troller BK importer

albertisfu opened the pull request:
In this PR I modified the `troller_bk` command in order to fix some details and added some changes so it can work as required in #2091 


The command will import data according to the following conditions:

- The importer won't modify the current data we already have.
- The importer won't add data after CL started scraping PACER RSS feeds.

The following tables describe when a docket or docket entries should be added, according to its current conditions and the RSS date.

**RSS before 2018-04-18**

  | District: | Appellate
-- | -- | --
**Docket exists, Has docket entries** | Don’t update docket metadata, Don’t add docket entries  | Don’t update docket metadata, Don’t add docket entries
**Docket exists, Empty docket no entries** | Add docket entries, Don’t update docket metadata | Add docket entries, Don’t update docket metadata
**Docket doesn’t exis**t | Add the new docket, Add  docket entries | Add the new docket, Add  docket entries



**RSS After 2018-04-18**

  | District: | Appellate
-- | -- | --
**Docket exists, Has docket entries** | Don’t update docket metadata , Don’t add docket entries | Don’t update docket metadata, Don’t add docket entries
**Docket exists, Empty docket no entries** | Don’t update docket metadata, Don’t add docket entries | Add docket entries, Don’t update docket metadata
**Docket doesn’t exist** | Don’t add docket, Don’t add docket entries | Add the new docket, Add  docket entries 

Added tests for each of these cases to confirm it works properly.

You can see in the second table **RSS After 2018-04-18** on the District column, we'll avoid adding anything after 2018-04-18 the date District and Bankruptcy RSS Feeds started being scraped by CL, so those cases are not supposed to happen since we should have all of these content, right?

Now I'm working on cleaning, merging, and ordering the S3 inventory files and saving them in a single one, I found the number after the `|` is the RSS date `o-894|1599853056` so `1599853056` is a Unix Timestamp, so I'll order the lines based on this field so we can do the import in chronological order, older firsts. 

Since we can know easily the date and court of RSS file, I was wondering if we should remove all the District and Bankruptcy lines after 2018-04-18?  so the processing is faster?

- Added a method to log on Redis the total dockets and entries added every 1000 `RECAPDocuments`, I set the key to has an expiration of 4 weeks,  so in case the importer fails when it's restarted the counter will continue adding the new values.

- I created the new branch because I had to solve some merge conflicts since your troller_bk branch was ahead of main specifically in pyproject.toml, and I set this PR to be merged on #2507 so the difs are not too long, you can see the actual changes in troller_bk command here: https://github.com/freelawproject/courtlistener/pull/2539/commits/00ac2717bb737f412c9099d474f71bed1b8624e9

Let me know what you think and I'll report here my findings after finishing cleaning and ordering the S3 inventory files.

mlissner commented:
Looks like your lint is getting hit by the mypy upgrade. I thought I had held that back so I could resolve it, but do you mind taking care of it, please? I think you've gotten better at mypy than I am, in any case. Hopefully it's easy stuff?

albertisfu commented:
Yes, of course, I'll solve the `mypy` issues!

mlissner reviewed (changes requested):
A couple tiny comments, but otherwise looks great.

mlissner commented:
A few replies:

> those cases are not supposed to happen since we should have all of these content, right?

Exactly, yes.

> I was wondering if we should remove all the District and Bankruptcy lines after 2018-04-18? so the processing is faster?

Yeah, it's worth doing. Our poor database will thank us.

> I set the key to has an expiration of 4 weeks

I appreciate that. It's good not to carry stats around forever.

Thank you!

albertisfu commented on cl/corpus_importer/management/commands/troller_bk.py:
```diff
@@ -109,38 +141,85 @@ class OptionsType(TypedDict):
         add_items_to_solr(rds_for_solr, "search.RECAPDocument")
 
+        line_counter += 1
+        logger.info(f"Last line imported: {line_counter}")
```
Yeah, you're right, the correct is to use `i`, changed, thanks!

albertisfu commented on cl/corpus_importer/management/commands/troller_bk.py:
```diff
@@ -109,38 +141,85 @@ class OptionsType(TypedDict):
+        line_counter += 1
+        logger.info(f"Last line imported: {line_counter}")
+        total_dockets_created = total_dockets_created + dockets_created
+        total_rds_created = total_rds_created + len(rds_for_solr)
```
Yeah, I changed it to `+=` and also solve some additional conflicts here, as described below.

albertisfu commented:
Thanks! I've applied the requested changes here, I just need to complete the `mypy` fixes.

Some additional tweaks:

It turned out that `log_added_items_to_redis` method was not logging the right number of dockets and rds, since it was adding the current value stored on redis to the total accumulated in an execution command. 

So in order to solve it, once the total accumulated is logged on Redis every 1000 lines, the total counters are restarted to 0, so the right value is added to the previous one on Redis.

I also changed it to log every 1000 lines instead of every 1000 rds created, since to log something was needed the new 1000 rds created matched exactly on the current line that's is pretty difficult.

- Changed the regexs in order to be able to parse every court in the list of paths, these after merging, cleaning, and ordering the inventory files.

I'm attaching here the Jupiter notebook that I used to merge, clean, order, and validate the inventory files:

[clean-and-order-troller-files.txt](https://github.com/freelawproject/courtlistener/files/10738284/clean-and-order-troller-files.txt)

I changed the extension to .txt in order to be able to attach it here, if you want to open it in a Jupiter notebook it would be enough to change the extension to `.ipynb` or I could change it to python if it's better.

Some comments about this process:

The total number of paths of all the inventory files is: `7,189,513`

From this total, there are `102,061` that are not RSS paths like:
`sources/troller-files/travis-park-pacer-data/district_pacer_doc_id/nysd-127010100215.pdf`

So after removing these there are `7,087,452` valid RSS paths.

Then I removed all the paths after 2018-04-18 that belong to a court we don't scrape, I used a script to get the courts in these paths that we don't scrape.
`['miwb', 'neb', 'nceb', 'pamd', 'cit', 'ca6', 'ca10', 'cadc', 'cafc']`

So any path after 2018-04-18 that doesn't belong to this list was removed (`3,971,393`)

After removing these paths remained `3,116,059` paths.

- One question about this is: did we start scraping all the courts we're now scraping at the same time after `2018-04-18`? Because if this is not true we might miss some paths that should be scrapped.

The merged files are:
[cleaned_and_ordered.csv](https://s3.console.aws.amazon.com/s3/object/com-courtlistener-private-storage?region=us-west-2&prefix=sources/troller-files/com-courtlistener-private-storage/troller-files/cleaned_and_ordered.csv) 
This one contains all the valid paths after its cleaning and orderer chronologically 

[final-file-cleaned-ordered.csv](https://s3.console.aws.amazon.com/s3/object/com-courtlistener-private-storage?region=us-west-2&prefix=sources/troller-files/com-courtlistener-private-storage/troller-files/final-file-cleaned-ordered.csv)
This one contains all the valid paths after removing paths from courts we scrape after 2018-04-18

[removed-lines-after-2018-04-18.csv](https://s3.console.aws.amazon.com/s3/object/com-courtlistener-private-storage?region=us-west-2&prefix=sources/troller-files/com-courtlistener-private-storage/troller-files/removed-lines-after-2018-04-18.csv)
This one contains the removed paths from courts we scrape after 2018-04-18

Let me know what you think.

mlissner commented:
That all sounds great, thanks.

> One question about this is: did we start scraping all the courts we're now scraping at the same time after 2018-04-18? Because if this is not true we might miss some paths that should be scrapped.

No, we didn't start at the same time for all courts, but it's close enough that I think we can just live with it. I don't think it'd be worth it to try to figure it out in any case.

mlissner commented:
Great. Let me know when things are ready for me. I've lost track of what depends on what.
